### PR TITLE
922 Implement static type switch on safe union

### LINKED ISSUE #922: Implement static switch based on type for safe union
**What Needs to be Done?**

Right now to switch on the types in a `vt::SafeUnion` one must write an if statement calling `u.is<T>()`, which is error prone if all cases must be properly covered. Implement a static functor dispatch across the union, which will simplify several use cases.

### PR DESCRIPTION
**Pull request by lifflander**
Fixes #922 

Another spin off of #882, that provides a lot more safety when switching.

### CODE REVIEW

File: tests/unit/utils/test_safe_union.cc [hunk 1]
```diff
@@ -226,5 +226,75 @@ TEST_F(TestSafeUnion, test_safe_union_3) {
...
+TEST_F(TestSafeUnion, test_safe_union_switch_4) {
+
+  vt::adt::SafeUnion<float, int, MyTest4> x;
+
+  EXPECT_EQ(x.template switchOn<TestFunctor>(x), 0);
+
+  x.init<int>();
+  x.get<int>() = 10;
+
+  EXPECT_EQ(x.template switchOn<TestFunctor>(x), 2);
```
Comment (PhilMiller) [inline tests/unit/utils/test_safe_union.cc:290]: This is a really funny looking interface. If it's a member function of `x`, why should the caller have to pass `x` again? Shouldn't the functor be taking exactly an argument of the selected type?

**Comment by PhilMiller** [comment]
I'm fine with approving this as is, if it's really what you want, but it seems like a very strange way to express it.

**Comment by lifflander** [comment]
> I'm fine with approving this as is, if it's really what you want, but it seems like a very strange way to express it.

@PhilMiller I've designed it so you can pass whatever arguments you want to the type-switched functor. In the actual use case, I have several args to pass. Unless I always pass as the first argument, I don't know how to easily "automatlly" pass the union, if that's what is needed, while keeping the arguments aligned in caller/callee.

If you have a better idea, I'm completely open to it.

**Comment by PhilMiller** [comment]
I think my instinct is to let the functor capture any other arguments it may need, and to pass the active value from the union as the only argument. That eliminates the variadic templating and the argument forwarding entirely.

**Comment by lifflander** [comment]
> I think my instinct is to let the functor capture any other arguments it may need, and to pass the active value from the union as the only argument. That eliminates the variadic templating and the argument forwarding entirely.

The functor can't be a lambda in C++14 since templated lambdas don't exist. Thus, there's no way to "capture". Am I misunderstanding what you are suggesting?

**Comment by PhilMiller** [comment]
[Generic lambdas](http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2013/n3649.html) is in C++14

**Comment by PhilMiller** [comment]
Or, the caller can actually define a functor, just like you did in the test, and construct it with suitable arguments

**Comment by lifflander** [comment]
I'm still missing how this would work. The functor is templated:
```C++
template <typename T>
struct MyFunctor {
  void operator()() const { }
};
```
The switch instantiates the functor with the right type. The user can't create one, because they don't know what type the union is on, and thus wouldn't be able to construct it. Are you suggesting that we move the template inside to the `operator()` and template it over `T`?

**Comment by PhilMiller** [comment]
Along those lines, the functor's `operator()` can either be a template, or an overload set. The union's switch would just need to call it with the appropriate argument type, and the compiler will do the right thing.

**Comment by lifflander** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- tests/unit/utils/test_safe_union.cc  6
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/lifflander/vt/pullRequest?prid=5845361&bid=19137955)

**Pull request merged**